programs/cltester/dwarf.cpp

```diff
@@ -928,8 +1176,221 @@ namespace dwarf
+                          std::vector<char>& symbol_data,
+                          const info& info,
+                          const std::vector<jit_fn_loc>& fn_locs,
+                          const std::vector<jit_instr_loc>& instr_locs,
+                          const void* code_start,
+                          size_t code_size)
+   {
+      std::map<die_pattern, uint32_t> codes;
+      die_pattern die;
+
+      eosio::vector_stream info_s{info_data};
+      auto begin_pos = info_data.size();
+      eosio::to_bin(uint32_t(0xffff'ffff), info_s);
+      auto length_pos = info_data.size();
+      eosio::to_bin(uint64_t(0), info_s);
+      auto inner_pos = info_data.size();
+
+      eosio::to_bin(uint16_t(compile_unit_version), info_s);
+      eosio::to_bin(uint64_t(0), info_s);  // debug_abbrev_offset
+      eosio::to_bin(uint8_t(8), info_s);   // address_size
+
+      die.tag = dw_tag_compile_unit;
+      die.has_children = true;
+      die.attrs = {
+          {dw_at_language, dw_form_data8, uint64_t(dw_lang_c_plus_plus)},
+          {dw_at_low_pc, dw_form_addr, uint64_t(code_start)},
+          {dw_at_high_pc, dw_form_addr, uint64_t((char*)code_start + code_size)},
+          {dw_at_stmt_list, dw_form_sec_offset, uint64_t(0)},
+      };
+      write_die(0, abbrev_data, info_data, codes, die);
+
+      Elf64_Sym null_sym;
+      memset(&null_sym, 0, sizeof(null_sym));
+      symbol_data.insert(symbol_data.end(), (char*)(&null_sym), (char*)(&null_sym + 1));
+
+      std::vector<size_t> sub_positions(info.subprograms.size());
+      std::vector<sub_ref> sub_refs;
+
+      for (size_t i = 0; i < info.subprograms.size(); ++i)
+      {
+         auto& sub = info.subprograms[i];
+         sub_positions[i] = info_data.size();
+         auto fn = get_wasm_fn(info, info.wasm_code_offset + sub.begin_address);
+         if (!fn || info.wasm_code_offset + sub.end_address > info.wasm_fns[*fn].end_pos)
+         {
+            if (show_generated_dies)
+               fprintf(stderr, "address lookup fail: %s %08x-%08x\n", sub.demangled_name.c_str(),
+                       info.wasm_code_offset + sub.begin_address,
+                       info.wasm_code_offset + sub.end_address);
+            continue;
+         }
+         if (sub.parent)
+            continue;
+         auto fn_begin = uint64_t((const char*)code_start + fn_locs[*fn].code_prologue);
+         auto fn_end = uint64_t((const char*)code_start + fn_locs[*fn].code_end);
+         if (show_generated_dies)
+            fprintf(stderr, "    DIE 0x%lx (%ld) subprogram %08x-%08x %016lx-%016lx %s\n",
+                    uint64_t(info_data.size()), uint64_t(i),
+                    info.wasm_code_offset + sub.begin_address,
+                    info.wasm_code_offset + sub.end_address, fn_begin + print_addr_adj,
+                    fn_end + print_addr_adj, sub.demangled_name.c_str());
+
+         die.tag = dw_tag_subprogram;
+         die.has_children = false;
+         if (sub.parent)
```

Comment:
Didn't we just check `sub.parent` and bail out above?

Reply:
At one point I had the code partially completed to generate a tree for inlined code. I stripped most of it out to finish later if needed (clang still inlines in spots even with `-O1`; `-fno-inline-functions` doesn't even prevent it). I'll remove this.

Reply:
removed